shared/utils/fix-cces.py

```diff
@@ -0,0 +1,168 @@
+                        break
+                    line_num += 1
+
+                end = line_num - 1
+                sec_ranges.append((begin, end))
+        line_num += 1
+    return sec_ranges
+
+def fix_empty_identifier(path, product_yaml):
+    file_contents = open(path, 'r').read().split("\n")
+    if file_contents[-1] == '':
+        file_contents = file_contents[:-1]
+
+    yaml_contents = ssgcommon.open_and_macro_expand_yaml(path, product_yaml)
+
+    empty_identifiers = []
+    for i_type, i_value in yaml_contents['identifiers'].items():
+        if i_value.strip() == "":
+            empty_identifiers.append(i_type)
+
+    print("====BEGIN BEFORE====")
+    print_file(file_contents)
+    print("====END BEFORE====")
+    sec_ranges = find_section_lines(file_contents, 'identifiers')
+    if len(sec_ranges) != 1:
+        raise RuntimeError("Refusing to fix file: %s -- could not find one section: %d" % (path, sec_ranges))
+
+    begin, end = sec_ranges[0]
+
+    # If num keys == num empty identifiers, then we can remove the entire
+    # identifiers section
+    if len(yaml_contents['identifiers'].keys()) == len(empty_identifiers):
+        file_contents = file_contents[:begin] + file_contents[end+1:]
+        print("Removing entire section since all keys are empty")
+    else:
+        raise RuntimeError("Refusing to fix file: %s -- additional keys in section" % path)
+
+    print("====BEGIN AFTER====")
+    print_file(file_contents)
+    print("====END AFTER====")
+    # response = raw_input("Confirm writing output to %s: (y/n): " % path)
+    response = 'y'
+    if response.strip() == 'y':
+        f = open(path, 'w')
+        for line in file_contents:
+            f.write(line)
+            f.write("\n")
+        f.flush()
+        f.close()
+
+def fix_empty_identifiers(directory):
+    results = find_rules(directory, has_empty_identifier)
+    print("Number of rules with empty identifiers: %d" % len(results))
+
+    for result in results:
+        rule_path = result[0]
+        product_yaml_path = result[1]
+
+        product_yaml = None
+        if product_yaml_path is not None:
+            product_yaml = ssgcommon.open_yaml(product_yaml_path)
+
+        fix_empty_identifier(rule_path, product_yaml)
+
+def find_invalid_cce(directory):
+    results = find_rules(directory, has_invalid_cce)
+    print("Number of rules with invalid CCEs: %d" % len(results))
+   
+ 
+    for result in results:
+        rule_path = result[0]
+        print(rule_path)
+
+def __main__():
+    if sys.argv[1] == 'empty':
+        fix_empty_identifiers(sys.argv[2])
+    elif sys.argv[1] == 'invalid':
+        find_invalid_cce(sys.argv[2])
+    else:
+        print("Usage: %s mode /full/path/to/src/directory" % sys.argv[0])
+        print("Modes:")
+        print("\tempty - check and fix rules with empty identifiers")
+        print("\tinvalid - find rules with invalid identifiers")
+
+if __name__ == "__main__":
+    __main__()
```

Comment:
hmm, this is a cool script and might help us write other scripts but on the other hand it is a one-off, I don't expect anyone to ever run it again because build system now forbids invalid identifiers so there is no way to introduce them